engine/runtime/src/test/java/org/enso/interpreter/test/JavaInteropTest.java

```diff
@@ -163,12 +165,96 @@ public void testImportNonExistingInnerNestedClass() {
   public void testImportOuterClassAndAccessNestedInnerClass() {
     var code = """
         polyglot java import org.enso.example.TestClass
-        
+
         main =
             instance = TestClass.StaticInnerClass.StaticInnerInnerClass.new
             instance.mul 3 5
         """;
     var res = evalModule(ctx, code);
     assertEquals(15, res.asInt());
   }
+
+  @Test
+  public void testToStringBehavior() {
+    var code = """
+    from Standard.Base import all
+
+    polyglot java import org.enso.example.ToString as Foo
+
+    type My_Fooable_Implementation
+        Instance x
+
+        foo : Integer
+        foo self = 100+self.x
+
+    main =
+        fooable = My_Fooable_Implementation.Instance 23
+        a = fooable.foo
+        b = fooable.to_text
+        c = Foo.callFoo fooable
+        d = Foo.showObject fooable
+        e = Foo.callFooAndShow fooable
+        [a, b, c, d, e]
+    """;
+
+    var res = evalModule(ctx, code);
+    assertTrue("It is an array", res.hasArrayElements());
+    assertEquals("Array with five elements", 5, res.getArraySize());
+    assertEquals(123, res.getArrayElement(0).asInt());
+    assertEquals("(Instance 23)", res.getArrayElement(1).asString());
+    assertEquals("Fooable.foo() = 123", res.getArrayElement(2).asString());
+    assertEquals("obj.toString() = (Instance 23)", res.getArrayElement(3).asString());
+    assertEquals("{(Instance 23)}.foo() = 123", res.getArrayElement(4).asString());
+  }
+
+  @Test
+  public void testInterfaceProxyFailuresA() {
+    var payload = evalInterfaceProxyFailures("a");
+    assertEquals("My_Exc", payload.getMetaObject().getMetaSimpleName());
+    var stdout = getStdOutLines();
+    var expectedLines = List.of("Executing Fooable_Panic.foo");
+    assertArrayEquals(expectedLines.toArray(), stdout);
+  }
+
+  @Test
+  public void testInterfaceProxyFailuresB() {
+    var result = evalInterfaceProxyFailures("b");
+    assertEquals("foo", result.asString());
+    var stdout = getStdOutLines();
+    var expectedLines = List.of("Executing Fooable_Unresolved.foo", "Executing Fooable_Unresolved.foo");
```

Comment:
Invoking the `b` function fails with wrong (`foo`) identifier and moreover it runs the `foo` code twice! The [conversion](https://github.com/enso-org/enso/pull/7487/files#r1284974772) needs to differentiate between:
- unknown symbol on the boundary
- unknown symbol during execution

Only the boundary symbol shall be converted to `InteropException`.